libs/ml/include/ml/optimisation/optimiser.hpp

```diff
@@ -124,13 +142,77 @@ typename T::Type Optimiser<T, C>::Run(ArrayType &data, ArrayType &labels, SizeTy
     // Compute and apply gradient
     ApplyGradients(batch_size);
 
-    FETCH_LOG_INFO("ML_LIB", "Loss: ", loss);
+    FETCH_LOG_INFO("ML_LIB", "Batch loss: ", loss);
 
     step += batch_size;
+    loss_sum += loss;
   }
 
-  DataType loss_sum{0};
-  loss_sum += loss;
+  epoch_++;
+
+  return loss_sum;
+}
+
+/**
+ * Does 1 training epoch using DataLoader
+ * @tparam T ArrayType
+ * @tparam C CriterionType
+ * @param loader DataLoader that provides examples for training
+ * @param batch_size size of mini-batch
+ * @param subset_size size of subset that will be loaded from DataLoader and trained in 1 epoch. If
+ * not specified, epoch will end when DataLoader wouldn't have next example. Loader is being reset
+ * on begining of each epoch.
+ * @return Sum of losses from all mini-batches
+ */
+template <class T, class C>
+typename T::Type Optimiser<T, C>::Run(fetch::ml::DataLoader<ArrayType, ArrayType> &loader,
+                                      bool random_mode, SizeType batch_size, SizeType subset_size)
+{
+  loader.Reset();
```

Comment:
I think replace loader.Reset() with
if (loader.IsDone())
{
  loader.Reset();
}

this way you can call Run with a subset size and increment through the training data (rather than always repeat on the first subset size.